src/blocktypes/containers.nix

```diff
@@ -22,92 +23,88 @@
+          local uri prev_tag
+          uri=$1
+          shift
+
+          for tag in $(<${tags'}); do
+            if ! [[ -v prev_tag ]]; then
+              ${skopeo} copy nix:${target} "$uri:tag" "$@"
+            else
+              # speedup: copy from the previous tag to avoid superflous network bandwidth
+              ${skopeo} copy "$uri:$prev_tag" "$uri:tag" "$@"
+            fi
+
+            prev_tag="$tag"
+          done
+        }
+      '';
+    in [
       (sharedActions.build currentSystem target)
       (mkCommand currentSystem {
         name = "print-image";
-        description = "print out the image name & tag";
+        description = "print out the image name with all tags";
         command = ''
           echo
-          echo "${target.imageName}:${target.imageTag}"
+          for tag in $(<${tags'}); do
+            echo "${img}:$tag"
+          done
         '';
       })
       (mkCommand currentSystem {
         name = "publish";
         description = "copy the image to its remote registry";
-        command = let
-          image = target.imageRefUnsafe or "${target.imageName}:${target.imageTag}";
-        in ''
-          # docker://${builtins.unsafeDiscardStringContext image}
-          ${target.copyToRegistry}/bin/copy-to-registry
+        command = ''
+          ${copyFn}
+          copy docker://${img}
         '';
-        proviso =
-          l.toFile "container-proviso"
-          # bash
-          ''
-            function proviso() {
-            local -n input=$1
-            local -n output=$2
-
-            local -a images
-            local delim="$RANDOM"
-
-            function get_images () {
-              command nix show-derivation $@ \
-              | command jq -r '.[].env.text' \
-              | command grep -o 'docker://\S*'
-            }
-
-            drvs="$(command jq -r '.actionDrv | select(. != "null")' <<< "''${input[@]}")"
-
-            mapfile -t images < <(get_images $drvs)
-
-            command cat << "$delim" > /tmp/check.sh
-            #!/usr/bin/env bash
-            if ! command skopeo inspect --insecure-policy "$1" &>/dev/null; then
-            echo "$1" >> /tmp/no_exist
-            fi
-            $delim
-
-            chmod +x /tmp/check.sh
-
-            rm -f /tmp/no_exist
-
-            echo "''${images[@]}" \
-            | command xargs -n 1 -P 0 /tmp/check.sh
-
-            declare -a filtered
-
-            for i in "''${!images[@]}"; do
-              if command grep "''${images[$i]}" /tmp/no_exist &>/dev/null; then
-                filtered+=("''${input[$i]}")
-              fi
-            done
-
-            output=$(command jq -cs '. += $p' --argjson p "$output" <<< "''${filtered[@]}")
-            }
-          '';
+        proviso = l.toFile "container-proviso" ''
+          SKOPEO_INSPECT=${./proviso/skopeo-inspect.sh}
+          ${builtins.readFile ./proviso/publish.sh}
+        '';
+        meta.tags = map (tag: "${img}:${tag}") tags;
```

Comment:
Maybe we also should name this something else? As it's not `tags`, anymore? It's probably correctly `images`?

Reply:
yeah I suppose that makes sense

Reply:
That would also clarify `tags[0]` to `images[0]` in the proviso.